openjdk/barriers/mmtkSATBBarrier.cpp

```diff
@@ -0,0 +1,536 @@
+    }
+    slot = ptr;
+  } else if (needs_patching && !slot->is_address()) {
+    assert(slot->is_register(), "must be");
+    slot = LIR_OprFact::address(new LIR_Address(slot, T_OBJECT));
+  }
+  assert(needs_patching || slot->is_register(), "must be a register at this point unless needs_patching");
+  if (!new_val->is_register()) {
+    LIR_Opr new_val_reg = gen->new_register(T_OBJECT);
+    if (new_val->is_constant()) {
+      __ move(new_val, new_val_reg);
+    } else {
+      __ leal(new_val, new_val_reg);
+    }
+    new_val = new_val_reg;
+  }
+  assert(new_val->is_register(), "must be a register at this point");
+  MMTkC1PreBarrierStub* slow = new MMTkC1PreBarrierStub(src, slot, new_val, info, needs_patching ? lir_patch_normal : lir_patch_none);
+  if (needs_patching) slow->scratch = gen->new_register(T_OBJECT);
+
+#if MMTK_ENABLE_BARRIER_FASTPATH
+  if (needs_patching) {
+    // At this stage, slot address is not available, so cannot do the fast-path check until 
+    // its address get resolved
+    // FIXME: Jump to a medium-path for code patching without entering slow-path
+    __ jump(slow);
+  } else {
+    // // load pre_val 
+    // LIR_Address* slot_addr = new LIR_Address(slot, T_OBJECT);
+    // LIR_Opr addr = slot;
+    // __ load(slot_addr, addr);
+    // // if pre_val == NULL skip the barrier
+    // __ cmp(lir_cond_equal, addr, LIR_OprFact::oopConst(NULL));
+    // __ branch(lir_cond_equal, T_OBJECT, slow->continuation());
+    LIR_Opr addr = src;
+    // uint8_t* meta_addr = (uint8_t*) (side_metadata_base_address() + (addr >> 6));
+    LIR_Opr offset = gen->new_pointer_register();
+    __ move(addr, offset);
+    __ unsigned_shift_right(offset, UseCompressedOops ? 5 : 6, offset);
+    LIR_Opr base = gen->new_pointer_register();
+    __ move(LIR_OprFact::longConst(side_metadata_base_address()), base);
+    LIR_Address* meta_addr = new LIR_Address(base, offset, T_BYTE);
+    // uint8_t byte_val = *meta_addr;
+    LIR_Opr byte_val = gen->new_register(T_INT);
+    __ move(meta_addr, byte_val);
+   
+    // intptr_t shift = (addr >> 3) & 0b111;
+    LIR_Opr shift = gen->new_register(T_INT);
+    __ move(addr, shift);
+    __ unsigned_shift_right(shift, UseCompressedOops ? 2 : 3, shift);
+    __ logical_and(shift, LIR_OprFact::intConst(0b111), shift);
+    // if (((byte_val >> shift) & 1) == 1) slow;
+    LIR_Opr result = byte_val;
+    __ unsigned_shift_right(result, shift, result, LIR_OprFact::illegalOpr);
+    __ logical_and(result, LIR_OprFact::intConst(1), result);
+    __ cmp(lir_cond_equal, result, LIR_OprFact::intConst(1));
+    __ branch(lir_cond_equal, T_BYTE, slow);
+  }
+#else
+  __ jump(slow);
+#endif
+
+  __ branch_destination(slow->continuation());
+}
+
+#undef __
+
+#define __ ideal.
+
+void MMTkSATBBarrierSetC2::object_reference_write_pre(GraphKit* kit, Node* src, Node* slot, Node* pre_val, Node* val) const {
+  if (can_remove_barrier(kit, &kit->gvn(), src, slot, val, /* skip_const_null */ false)) return;
```

Comment:
Be careful of barrier elimination when implementing this in the binding. target == null does not mean the barrier can be eliminated. This bug takes me a while to find out.